raster-foundry/raster-foundry
# Add cascade deletion when it makes sense

## Overview

This PR adds cascade deletion (and one `SET NULL`) for:
 - [x] `projects` -> `annotation_groups`
 - [x] `projects` -> `annotations`
 - [x] `projects` -> `aois`
 - [x] `projects` -> `exports`
 - [x] `annotation_groups` -> `annotations`
 - [x] `tool_runs` -> `exports`
 - [x] `feature_flags` -> `organization_features`
 - [x] `platforms` -> `organizations`
 - [x]  `organizations` -> `organization_features`
 - [x] `organizations` -> `teams`
 - [x] `licenses` -> `datasource`'s `license_name` column `SET NULL`

It also updates a property test in `PlatformDaoSpec`, which fails sporadically, to fully reflect the use of `searchQP`.


### Checklist

- ~Styleguide updated, if necessary~
- ~[Swagger specification](https://github.com/raster-foundry/raster-foundry-api-spec) updated~
- ~Any content changes are properly templated using `BUILDCONFIG.APP_NAME`~
- ~Any new SQL strings have tests~

### Notes

 - This PR does not include cascade deletions of `datasources` -> `scenes` and `datasources` -> `uploads`, since users should be forced to select or create another datasource, which will update the datasource column in the corresponding `scenes` and `uploads` (per discussions in https://github.com/raster-foundry/raster-foundry/issues/3111).
 - I think we should probably add some messages in the deletion modal alerting that referenced data will be deleted as well (like when deleting a project with annotations, we should probably say something in the deletion modal like "Your annotations will be deleted along with this project").
 - We have some cascade deletions already, including the mentioned `Scenes -> Images -> Bands` and `Scenes -> Thumbnails` in the original issue.


## Testing Instructions

 * Check if the proposed cascade deletions make sense
 * Check if any cascade deletion is left behind (except the two mentioned in notes section)
 * Run migrations
 * Check if deleting a record in the referenced table results in the deletion of a record in the referencing table.
 * CI

Closes #2288 
Closes #3690
